File: cps-444/homework-3/cmd.go
package main

import (
	"fmt"
	"os"
	"strings"

	"github.com/spf13/pflag"
)

// Create the set of flags for our wc command.
var fs = pflag.NewFlagSet("wc", pflag.ContinueOnError)
var lines = fs.BoolP("lines", "l", false, "print the newline counts")
var words = fs.BoolP("words", "w", false, "print the word counts")
var chars = fs.BoolP("chars", "m", false, "print the character counts")
var helpFlag = fs.Bool("help", false, "display this help and exit")

func init() {
	// Any flag-related error messages will be printed to stderr.
	fs.SetOutput(os.Stderr)

	fs.Usage = func() {
		if f := fs.Lookup("help"); f != nil {
			f.Hidden = true
		}
		var msg = `Usage: wc [OPTION]... [FILE]...
  or:  wc [OPTION]... --files0-from=F
Print newline, word, and byte counts for each FILE, and a total line if
more than one FILE is specified.  A word is a non-zero-length sequence of
printable characters delimited by white space.

With no FILE, or when FILE is -, read standard input.

The options below may be used to select which counts are printed, always in
the following order: newline, word, character, byte, maximum line length.`
		// The help message is printed to stdout.
		fs.SetOutput(os.Stdout)
		fmt.Println(msg)
		fs.PrintDefaults()
		fmt.Println("\n  --help     display this help and exit")
	}
}

func Execute() error {
	// If there is some sort of invalid option then print an error message and return early.
	err := fs.Parse(os.Args[1:])
	if err != nil {
		return checkInvalidOptions(err)
	}

	// Print the help message if the flag was passed.
	if *helpFlag {
		fs.Usage()
		return nil
	}

	fmt.Printf("%t %t %t\n", *lines, *words, *chars)
	return nil
}

func checkInvalidOptions(err error) error {
	var invalidOption string
	// If the erroneous flag is a shorthand flag (i.e. a one-letter abbreviation)
	// then the error will contain "shorthand flag". If the erroneous flag is a fully
	// described flag then the error will contain "unknown flag". We can use these
	// to get the error messages exactly right.
	if strings.Contains(err.Error(), "shorthand flag") {
		// An example error message might look something like this:
		// "unknown shorthand flag: 'a' in -asef"
		start := strings.Index(err.Error(), "'")
		if start != -1 {
			end := strings.Index(err.Error()[start+1:], "'")
			if end != -1 {
				invalidOption = err.Error()[start+1 : start+1+end]
			}
		}
		// For some reason wc has different error messages for shorthand and full-length flags.
		// Here we print "invalid option", down farther we use "unrecognized option".
		if invalidOption != "" {
			fmt.Fprintf(os.Stderr, "wc: invalid option -- '%s'\n", invalidOption)
		}
	} else if strings.Contains(err.Error(), "unknown flag") {
		// An example error message might look like this:
		// "unknown flag: --wawmlsefasef", which is much easier to parse.
		parts := strings.Fields(err.Error())
		if len(parts) > 0 {
			flagPart := parts[len(parts)-1]
			if strings.HasPrefix(flagPart, "--") {
				invalidOption = flagPart
			}
		}
		// Here is where we use "unrecognized option" as the error message to match the real command.
		if invalidOption != "" {
			fmt.Fprintf(os.Stderr, "wc: unrecognized option '%s'\n", invalidOption)
		}
	}
	// If we weren't able to figure out what the invalid option then print the whole error.
	// Hopefully we can modify the program to parse it correctly.
	if invalidOption == "" {
		fmt.Fprintf(os.Stderr, "wc: %v\n", err)
	}
	// End with a help message.
	fmt.Fprintf(os.Stderr, "Try 'wc --help' for more information.\n")
	return fmt.Errorf("invalid option")
}

File: cps-444/honors-homework-2/cmd.c
#include <getopt.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "mydefs.h"

// Define the static help message.
static char *help_msg = "Usage: wc [OPTION]... [FILE]...\n"
                        "  or:  wc [OPTION]... --files0-from=F\n"
                        "Print newline, word, and byte counts for each FILE, and a total line if\n"
                        "more than one FILE is specified.  A word is a non-zero-length sequence of\n"
                        "printable characters delimited by white space.\n"
                        "\n"
                        "With no FILE, or when FILE is -, read standard input.\n"
                        "\n"
                        "The options below may be used to select which counts are printed, always in\n"
                        "the following order: newline, word, character, byte, maximum line length.\n"
                        "  -m, --chars   print the character counts\n"
                        "  -l, --lines   print the newline counts\n"
                        "  -w, --words   print the word counts\n"
                        "\n"
                        "  --help     display this help and exit";

int parse_cmd(int argc, char *argv[], struct wc_stat_flags *wc_flags, int *first_pos_arg_ind)
{
    static struct option long_options[] = {
        {"lines", no_argument, NULL, 'l'}, // return 'l' in switch case
        {"words", no_argument, NULL, 'w'}, // return 'w'
        {"chars", no_argument, NULL, 'm'}, // return 'm'
        {"help", no_argument, NULL, 'h'},  // return 'h'
        {0, 0, 0, 0},                      // required array terminator
    };

    // The short_options string matches the shorthand option values for the long_options array.
    const char *short_options = "lwmh";

    // Loop through all the options and parse them one by one.
    // Setting opterr=0 allows for full control of error messages for invalid options.
    opterr = 0;
    bool was_stat_option_given = false;
    while (1)
    {
        // Store the index of the long option that was just found.
        int option_index = 0;
        // Parse out the next option.
        int curr_opt = getopt_long(argc, argv, short_options, long_options, &option_index);

        // This is the sentinel value for the end of the options.
        if (curr_opt == -1)
            break;

        // Handlers for each short option.
        switch (curr_opt)
        {
        // Parse out the different stat flags.
        case 'l':
            wc_flags->newlines = true;
            was_stat_option_given = true;
            break;
        case 'w':
            wc_flags->words = true;
            was_stat_option_given = true;
            break;
        case 'm':
            wc_flags->bytes = true;
            was_stat_option_given = true;
            break;

        // Print help message.
        case 'h':
            puts(help_msg);
            exit(EXIT_SUCCESS);

        // Handle unknown options.
        case '?':
            // Check if the problematic argument started with "--" to identify it as a long option error.
            if (argv[optind - 1] != NULL && strncmp(argv[optind - 1], "--", 2) == 0)
            {
                fprintf(stderr, "wc: unrecognized option '%s'\n", argv[optind - 1]);
            }
            // Otherwise, check for a short option error.
            else if (optopt != 0)
            {
                fprintf(stderr, "wc: invalid option -- '%c'\n", optopt);
            }
            else
            {
                // If we get here then there was some sort of parsing error.
                fprintf(stderr, "wc: unknown or invalid argument.\n");
            }

            // Handles unknown options or missing arguments.
            fprintf(stderr, "Try 'wc --help' for more information.\n");
            return EXIT_FAILURE;

        default:
            // We should NEVER get here.
            abort();
        }
    }

    // If no stat options were given then the default is to show everything.
    if (!was_stat_option_given)
    {
        wc_flags->newlines = true;
        wc_flags->words = true;
        wc_flags->bytes = true;
    }

    // This variable will hold the starting index of the positional arguments
    // after all of the options have been parsed out.
    *first_pos_arg_ind = optind < argc ? optind : -1;

    return EXIT_SUCCESS;
}
